### PR TITLE
🚧 Preview url

### PR DESCRIPTION
**Pull request by sparkoo**
### What does this PR do?
adds runtime command attribute with preview url, when able to find matching component with exposed endpoint.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/13945

#### Docs PR
TODO: 
- devfile schema https://github.com/redhat-developer/devfile

### CODE REVIEW

File: wsmaster/che-core-sql-schema/src/main/resources/che-schema/7.3.0/1__add_preview_url_to_devfile_command.sql [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+--
+-- SPDX-License-Identifier: EPL-2.0
+--
+-- Contributors:
+--   Red Hat, Inc. - initial API and implementation
+--
+
+
+ALTER TABLE devfile_command ADD COLUMN preview_url_port INTEGER;
+ALTER TABLE devfile_command ADD COLUMN preview_url_path VARCHAR(255);
```
Comment (skabashnyuk) [inline wsmaster/che-core-sql-schema/src/main/resources/che-schema/7.3.0/1__add_preview_url_to_devfile_command.sql]: may be TEXT?

Comment (sparkoo) [reply wsmaster/che-core-sql-schema/src/main/resources/che-schema/7.3.0/1__add_preview_url_to_devfile_command.sql]: yes

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceService.java [hunk 2]
```diff
@@ -952,11 +963,87 @@ private WorkspaceDto asDtoWithLinksAndToken(WorkspaceImpl workspace) throws Serv
...
         throw new ServerException(e.getMessage(), e);
       }
+
+      fixRuntimeCommandsPreviewUrls(workspaceDto);
     }
 
     return workspaceDto;
   }
 
+  private void fixRuntimeCommandsPreviewUrls(WorkspaceDto workspaceDto) {
```
Comment (mshaposhnik) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceService.java:973]: I think this whole logic deserves to be in separate class (if it doesn't tied with service too much)

**Comment by mshaposhnik** [comment]
Generally LGTM, please do not forget to update docs/examples with this cool new feature  (BTW i don't know which doc is actual for devfile right now, @skabashnyuk  did you?)

**Comment by sparkoo** [comment]
> Generally LGTM, please do not forget to update docs/examples with this cool new feature (BTW i don't know which doc is actual for devfile right now, @skabashnyuk did you?)

I know only about this https://github.com/redhat-developer/devfile

**Comment by mshaposhnik** [comment]
AFAIK it is not supported anymore

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceService.java [hunk 3]
```diff
@@ -952,11 +963,87 @@ private WorkspaceDto asDtoWithLinksAndToken(WorkspaceImpl workspace) throws Serv
       } catch (MachineTokenException e) {
         throw new ServerException(e.getMessage(), e);
       }
+
+      fixRuntimeCommandsPreviewUrls(workspaceDto);
```
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceService.java:967]: I believe this logic should not be performed on WorkspaceService layer.
Performing it once at workspace start seems right way for me https://github.com/eclipse/che/pull/14713/files

**Comment by skabashnyuk** [comment]
> I know only about this https://github.com/redhat-developer/devfile

deprecated
Please use https://www.eclipse.org/che/docs/che-7/configuring-a-workspace-using-a-devfile/ as starting point.

**Comment by sleshchenko** [comment]
Something that happened privately, did not happen at all =)
I mean there should be some deprecation process and it seems not to be done:
1. Migrate existing content https://redhat-developer.github.io/devfile/. I think information about ephemeral mode, default editor and plugins, brief components overview is useful.
2. Update that documentation last time with information that it's deprecated, more somenewurl instead.
@skabashnyuk Maybe we should create a dedicated issue for that?

**Comment by sparkoo** [comment]
yes, definitely strange. This is first time I hear about deprecation of that doc. It's even linked from https://www.eclipse.org/che/docs/che-7/configuring-a-workspace-using-a-devfile/.
I don't know how others, but I see the devfile specification very useful as I don't know other place where it is specified in human readable form (except the code). Also, we should not throw away one thing, if we don't have at least equivalent replace. And che-docs document is far from being complete or useful.

**Comment by sparkoo** [comment]
closing this one in favor of https://github.com/eclipse/che/pull/14713

**Pull request closed without merging**